### PR TITLE
[POAE7-2665] [WIP] HashJoinTranslator support

### PR DESCRIPTION
**Pull request by qiuyangshen**
### What changes were proposed in this pull request?

-   add HashJoinNode(Translator) and related RuntimeFuncitons, modify the parser to create HashJoinNode and add it to transform pipeline
-   add HashJoinTest, Construct a basic hash join sql to run the workflow to prove correctness

### Why are the changes needed?
For HashJoin codegen in Next Gen FW design

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UTs

### Which label does this PR belong to?
INFRA

### CODE REVIEW

File: cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp [hunk 1]
```diff
@@ -263,6 +264,45 @@ JITValuePointer LLVMJITFunction::getArgument(size_t index) {
...
+  // record memory index address
+  std::vector<int64_t> memory_index;
+  for (auto val : vals) {
+    memory_index.push_back(memory_count);
+    memory_count += getJITTypeSize(val->getValueTypeTag());
+    // memory align
+    memory_count = (memory_count + alignment) & ~(alignment - 1);
+  }
+
+  llvm::AllocaInst* allocated_memory = ir_builder_->CreateAlloca(
```
Comment (jikunshang) [inline cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp:280]: memory leak risk?

**Comment by qiuyangshen** [comment]
Sorry, now HashJoinTest is just an initial version，the copy from NextGenCompilerTest. I will modify it after integretion with Xieqi and Xinyi's HashTable design later.

File: cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp (hunk 1 above)
Comment (qiuyangshen) [reply cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp:280]: Yeah，this memory is allocated on the stack，so there is no such risk

File: cider/exec/nextgen/parsers/RelAlgEUParser.cpp [hunk 2]
```diff
@@ -116,6 +120,31 @@ static void insertSourceNode(const RelAlgExecutionUnit& eu, OpPipeline& pipeline
...
+  ExprPtrVector left_exprs, right_exprs;
+  auto input_exprs = analyzer.getInputExprs();
+  for (size_t i = 0; i < input_exprs.size(); ++i) {
+    if (nullptr == input_exprs[i]) {
+      LOG(FATAL) << "Input column expr missed. Column descriptor = "
+                 << analyzer.getInputDesc()[i].toString();
+    }
+    auto input = static_cast<Analyzer::ColumnVar*>(input_exprs[i].get());
+    // TODO:: how to differentiate build and probe table
+    input->get_table_id() == 0 ? left_exprs.push_back(input_exprs[i])
```
[the remark this answers is not in the document]
Comment (qiuyangshen) [reply cider/exec/nextgen/parsers/RelAlgEUParser.cpp:140]: Roger,  this TODO is about how to differentiate left and right table

File: cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp (hunk 1 above)
Comment (jikunshang) [reply cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp:280]: I guess you may pack each row in join scenario. If so, stack may overflow?

Comment (qiuyangshen) [reply cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp:280]: Actually I will use to pack mutiple column data in one row，the stack may be 8MB i'm not sure. So it is unlikely that this overflow. And I've push these code in this PR #226

File: cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp [hunk 3]
```diff
@@ -263,6 +264,45 @@ JITValuePointer LLVMJITFunction::getArgument(size_t index) {
   }
 }
 
+JITValuePointer LLVMJITFunction::packJITValuesImpl(
```
Comment (spevenhe) [inline cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp:267]: This function is used for transfer ```vector``` to ```array```?

Comment (qiuyangshen) [reply cider/exec/nextgen/jitlib/llvmjit/LLVMJITFunction.cpp:267]: I'm afraid not, this function is just used to pack Join keys for call the loopUpbyKey function

**Comment by qiuyangshen** [comment]
I have new a pr , and handled all of the comments above, you can take a review there. This pr will be closed , thanks for the review.

**Pull request closed without merging**